File: src/bridge/messages.ts
import { invoke } from '@tauri-apps/api/core';

type BridgeId = string | number;

const toBridgeId = (value: BridgeId): string => String(value);

export interface EncryptedMessageRow {
    message_id: string;
    room_id: string;
    sender_id: string;
    encrypted_content: number[];
    message_type: string;
    spk_key_id?: number;
    otpk_key_id?: number;
    server_timestamp: number;
    received_at: number;
}

export interface DecryptedMessageRow {
    message_id: string;
    room_id: string;
    sender_id: string;
    plaintext: number[];
    content_type: string;
    message_timestamp: number;
    reply_to_id: string | null;
    is_edited: boolean;
    is_deleted: boolean;
}

export interface StoreEncryptedMessageInput {
    messageId: BridgeId;
    roomId: BridgeId;
    senderId: BridgeId;
    encryptedContent: number[];
    messageType: string;
    spkKeyId?: number;
    otpkKeyId?: number;
    serverTimestamp: number;
}

export interface StoreDecryptedMessageInput {
    userId: BridgeId;
    messageId: BridgeId;
    roomId: BridgeId;
    senderId: BridgeId;
    plaintext: number[];
    contentType: string;
    messageTimestamp: number;
    replyToId?: string;
    isEdited: boolean;
    isDeleted: boolean;
}

export const storeEncryptedMessage = ({
    messageId,
    roomId,
    senderId,
    encryptedContent,
    messageType,
    spkKeyId,
    otpkKeyId,
    serverTimestamp,
}: StoreEncryptedMessageInput): Promise<void> =>
    invoke('store_encrypted_message', {
        messageId: toBridgeId(messageId),
        roomId: toBridgeId(roomId),
        senderId: toBridgeId(senderId),
        encryptedContent,
        messageType,
        ...(spkKeyId !== undefined && { spkKeyId }),
        ...(otpkKeyId !== undefined && { otpkKeyId }),
        serverTimestamp,
    });

export const getEncryptedMessages = (
    roomId: BridgeId,
    limit: number,
    beforeTimestamp?: number,
): Promise<EncryptedMessageRow[]> =>
    invoke<EncryptedMessageRow[]>('get_encrypted_messages', {
        roomId: toBridgeId(roomId),
        limit,
        ...(beforeTimestamp !== undefined && { beforeTimestamp }),
    });

export const storeDecryptedMessage = ({
    userId,
    messageId,
    roomId,
    senderId,
    plaintext,
    contentType,
    messageTimestamp,
    replyToId,
    isEdited,
    isDeleted,
}: StoreDecryptedMessageInput): Promise<void> =>
    invoke('store_decrypted_message', {
        userId: toBridgeId(userId),
        messageId: toBridgeId(messageId),
        roomId: toBridgeId(roomId),
        senderId: toBridgeId(senderId),
        plaintext,
        contentType,
        messageTimestamp,
        replyToId: replyToId ?? null,
        isEdited,
        isDeleted,
    });

export const getDecryptedMessages = (
    userId: BridgeId,
    roomId: BridgeId,
    limit: number,
    beforeTimestamp?: number,
): Promise<DecryptedMessageRow[]> =>
    invoke<DecryptedMessageRow[]>('get_decrypted_messages', {
        userId: toBridgeId(userId),
        roomId: toBridgeId(roomId),
        limit,
        ...(beforeTimestamp !== undefined && { beforeTimestamp }),
    });

File: src-tauri/src/commands/messages.rs
//! # Message Commands
//!
//! Tauri commands that proxy message storage to the `store::message_store` module.
//! No business logic lives here — commands are thin adapters between the TypeScript
//! frontend and the SQLite persistence layer.

use crate::commands::core::{
    get_decrypted_messages_core, get_encrypted_messages_core, store_decrypted_message_core,
    store_encrypted_message_core,
};
use crate::store::db::{get_or_create_master_key, open_db};
use crate::store::message_store::{DecryptedMessageRow, EncryptedMessageRow};

// ── Encrypted messages ────────────────────────────────────────────

/// Persist a raw E2EE ciphertext blob pulled from the server.
/// Idempotent — duplicate `message_id` inserts are silently ignored.
#[tauri::command]
pub async fn store_encrypted_message(
    app: tauri::AppHandle,
    message_id: String,
    room_id: String,
    sender_id: String,
    encrypted_content: Vec<u8>,
    message_type: String,
    spk_key_id: Option<u32>,
    otpk_key_id: Option<u32>,
    server_timestamp: i64,
) -> Result<(), String> {
    let conn = open_db(&app)?;
    store_encrypted_message_core(
        &conn,
        &message_id,
        &room_id,
        &sender_id,
        encrypted_content,
        &message_type,
        spk_key_id,
        otpk_key_id,
        server_timestamp,
    )
}

/// Fetch encrypted messages for `room_id`, newest first.
/// Pass `before_timestamp` to paginate older pages.
#[tauri::command]
pub async fn get_encrypted_messages(
    app: tauri::AppHandle,
    room_id: String,
    limit: u32,
    before_timestamp: Option<i64>,
) -> Result<Vec<EncryptedMessageRow>, String> {
    let conn = open_db(&app)?;
    get_encrypted_messages_core(&conn, &room_id, limit, before_timestamp)
}

// ── Decrypted messages ────────────────────────────────────────────

/// Store a decrypted plaintext, re-encrypted with the user's local master key.
#[tauri::command]
pub async fn store_decrypted_message(
    app: tauri::AppHandle,
    user_id: String,
    message_id: String,
    room_id: String,
    sender_id: String,
    plaintext: Vec<u8>,
    content_type: String,
    message_timestamp: i64,
    reply_to_id: Option<String>,
    is_edited: bool,
    is_deleted: bool,
) -> Result<(), String> {
    let conn = open_db(&app)?;
    let key = get_or_create_master_key(&user_id)?;
    store_decrypted_message_core(
        &conn,
        &key,
        &user_id,
        &message_id,
        &room_id,
        &sender_id,
        plaintext,
        &content_type,
        message_timestamp,
        reply_to_id.as_deref(),
        is_edited,
        is_deleted,
    )
}

/// Fetch decrypted messages for `(user_id, room_id)`, newest first.
/// Pass `before_timestamp` to paginate older pages.
#[tauri::command]
pub async fn get_decrypted_messages(
    app: tauri::AppHandle,
    user_id: String,
    room_id: String,
    limit: u32,
    before_timestamp: Option<i64>,
) -> Result<Vec<DecryptedMessageRow>, String> {
    let conn = open_db(&app)?;
    let key = get_or_create_master_key(&user_id)?;
    get_decrypted_messages_core(&conn, &key, &user_id, &room_id, limit, before_timestamp)
}

// ── Tests ─────────────────────────────────────────────────────────

#[cfg(test)]
#[path = "tests/message_tests.rs"]
mod tests;
